Update intel-tbb to 2018.3 and improve version URLs, dependencies and patching.

chissg opened the pull request:
Improve to version -> URL translation to be algorithmic rather than manual.

Make binutils build dependency explicit where appropriate because we're sensitive to the assembler used.

Build's use of '-mrtm' is only patched away if it needs to be because the assembler is too old.

adamjstewart commented on var/spack/repos/builtin/packages/intel-tbb/package.py:
```diff
@@ -81,7 +93,8 @@ def coerce_to_spack(self, tbb_build_subdir):
         if spec.satisfies('%gcc@6.1:') and spec.satisfies('@:4.4.3'):
-            raise InstallError('Only TBB 4.4.4 and above build with GCC 6.1!')
+            raise \
+                InstallError('Only TBB 4.4.4 and above build with GCC >= 6.1!')
```
Can you actually convert this to a `conflicts` statement? That will cause Spack to crash during concretization instead of installation. It would look something like:
```python
conflicts('%gcc@6.1:', when='@:4.4.3', msg='Only TBB 4.4.4 and above build with GCC >= 6.1!')
```

adamjstewart commented on var/spack/repos/builtin/packages/intel-tbb/package.py:
```diff
@@ -34,33 +34,45 @@ class IntelTbb(Package):
     provides('tbb')
 
+    # Assembler version is important.
+    depends_on('binutils', type='build', when='platform=linux %gcc@4.8.0:')
```
See http://spack.readthedocs.io/en/latest/getting_started.html?highlight=binutils#binutils

chissg replied:
Unsure quite what to do here based on advice at link. The conditionally-applied patch mitigates the problem with older `binutils`, but I either apply it unconditionally (previous behavior) with efficiency implications, or have a version of `binutils` against which to compare to ascertain whether the patch should be applied or not. I could just remove the patch entirely, but then how does one let the user know what to do in case of a build failure (e.g. on Centos6-based systems)? Please advise.

gartung commented:
@scheibelp

chissg commented:
After some help from @scheibelp this PR has been refreshed, and hopefully addresses all concerns, including the `binutils`-related ones. If the compiler used is a modern-enough version of GCC and the OS is a variant of RHEL6, the RTM activation is disabled by patch.
